### DISCUSSION TITLE
SvelteKit application over `file://` protocol

### DISCUSSION THREAD

**Discussion by jerivas** [sveltejs/kit#12206] [Q&A]
Hello! For reasons outside of our control, we require our SvelteKit app to work over the `file://` protocol. The goal is to build the app, transfer files to client machines, and have them use it by launching `index.html`.

The biggest obstacle so far is that [browsers refuse to load ESM scripts over `file://` URLs](https://barker.codes/blog/javascript-modules-and-the-file-uri-scheme/), which means all JS is being blocked. It looks like Vite can be configured to target `es5`, but that also would remove modern JS syntax like `let` and `const`, which is not desired. I also looked into legacy builds, but support for them is currently an [open issue](https://github.com/sveltejs/kit/issues/12).

I have prepared a [demo repo](https://github.com/jerivas/sveltkit-file) configured to build the Svelte Kit demo app with the static adapter. You can try it yourself by building it and pointing your browser directly to `build/index.html`. You will get a blank screen and errors in the console:

![image](https://github.com/sveltejs/kit/assets/2391102/aaa687d7-a11c-4a99-9247-faf636a766b8)

**Comment by genaray** [comment]
Any Solution to this? Having the same problem rn... would be glad for any help :(

**Comment by brunnerh** [comment]
In newer Kit versions you can set `output.bundleStrategy` to `'inline'` ([output docs](https://svelte.dev/docs/kit/configuration#output)) and remove Vite's [`build.assetsInlineLimit`](https://vite.dev/config/build-options.html#build-assetsinlinelimit) by setting it to a large value/`Infinity`.

To make routing work, set the `router.type` to `hash` ([router docs](https://svelte.dev/docs/kit/configuration#router)). The links in the application will accordingly have to use `#/route/here`.

**Reply by genaray** [reply]
Thanks! Is it normal that it still generates multiple Files? It generates an Index along the typically App folder with .json in it. Looks wrong hmm

**Reply by brunnerh** [reply]
[Here would be a demo project](https://github.com/Rich-Harris/snek).

Are you using `@sveltejs/adapter-static`?

(There will always be multiple intermediate files somewhere which then are turned into the final application by the adapter.)

**Reply by genaray** [reply]
Thanks i will Look at it! Yes im using Adapter-Static, inline should still Work then? How do i create the final HTML tho? npm run build/vite build simply generates that folder with those intermediate files and the index.html. Loading that one results in an cors error tho, any Idea? ^^  (Im using shadcn in it, could this cause Problems?)

**Reply by brunnerh** [reply]
The build command should output a standalone file (with possibly a few non-essential files).
The intermediate files should all be somewhere in the `.svelte-kit` directory.

It looks like something is off in your case. Maybe the library has something to do with it, not sure.

Ideally create a minimal reproduction and ask a separate question, providing all the necessary files.

**Reply by Unknown78** [reply]
Is it SSG or SPA? Is SSG possible?

**Reply by brunnerh** [reply]
It's fully static, see e.g. the linked example further up.

**Reply by Unknown78** [reply]
https://svelte.dev/docs/kit/configuration#router
> What type of client-side router to use.
> 
> 'pathname' is the default and means the current URL pathname determines the route
> 'hash' means the route is determined by location.hash. In this case, SSR and **prerendering are disabled**. ...

---

Does that mean that it is SPA?
I don't understand the rationale behind the use of `hash` router type instead of the usual `pathname`.

**Reply by brunnerh** [reply]
It effectively is an SPA, though most SPAs would still pull data from some server instead of rendering it into the page on the server. Here the entire static application is in a single file, which requires hash-based routing.

I don't know if the inline bundle strategy can work without this, i.e. having it generate _multiple_ independent HTML files that inline their own JS & CSS. I suspect that this is not possible, though.
